KAFKA-2211: Adding simpleAclAuthorizer implementation and test cases.

Parth-Brahmbhatt opened the pull request.

Parth-Brahmbhatt committed:
KAFKA-2211: Adding simpleAclAuthorizer implementation and test cases.

Parth-Brahmbhatt commented:
The only test failure reported by the bot was 
kafka.producer.ProducerTest > testSendWithDeadBroker FAILED
    java.lang.AssertionError: Message set should have 1 message
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at kafka.producer.ProducerTest.testSendWithDeadBroker(ProducerTest.scala:260)

Which passes locally and seems very unrelated to the changes i made. I will try and run the test in a loop to see if I can reproduce.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+    inReadLock(lock) {
+      if (aclCache.contains(resource))
+        return aclCache.get(resource).get
+    }
```
It seems that it will be cleaner if we split this into getAcls and getAclsFromZK. The former will just read from the cache and the latter always read from ZK. Only the ZK listener will use the latter.

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  def updateAclChangedFlag(resource: Resource): Unit = {
+      ZkUtils.updatePersistentPath(zkClient, SimpleAclAuthorizer.AclChangedZkPath, resource.toString)
```
Hmm, this may not work well if two ACL changes happen quickly. The second one will override the the resource of the previous one. If a broker hasn't finished processing the previous change, it will miss it after the override. I thought we wanted to use the same approach as in ConfigChangeListener where the updates are written as a sequential ZK node.

Parth-Brahmbhatt replied:
I am confused. I thought you said a watch notification will never be missed or it will only be missed when a reconnection happens which is handled by zkClient,  which is why we don't need the sync thread. 

If client-A updated the /kafka-acl-chaged and set its data to "topic-1" (this should go throught zkQuorum) and then client-B updated the /kafka-acl-changed and set its data to "topic-2" are you saying we will only get notification for "topic-2" depending on how fast the change occurs? Doesn't that violate the first guarantee which led us to delete the hourly sync? 

I would like to point out that in the watch notification we are just relying on data sent as part of notification and not really reading /acl-changed one more time.

junrao replied:
Parth,

ZK watchers are actually one-timer watchers (http://zookeeper.apache.org/doc/r3.4.1/zookeeperProgrammers.html#sc_WatchRememberThese). When a watcher fires, the client has to register the watcher again (typically through a read) to get notification for future changes. Between the time that a watcher fires and the client does a read, multiple changes could have happened. The read will return the latest value and leave a watcher there. So, in general, there is no guarantee that every change triggers the firing of a watcher. In particular, if you update the value of a ZK path multiple times in between, only the latest value will be seen by the reader.

One way to capture all changes is to follow how we propagate the config changes (ConfigChangeListener). Whenever we change an ACL, we write the ACL under /kafka-acl and also write a sequential ZK node under /kafka-acl-changed. The value of the sequential ZK node indicates the resource for which the ACL has changed. The sequential ZK nodes are guaranteed to be unique and have a number suffix that's ever growing. The ACL manager registers a child change listener on /kafka-acl-changed and remembers the last number suffix that it has processed. When the watcher fires, the ACL manager reads all child nodes under /kafka-acl-changed (the read may pick up multiple ACL changes) and process new nodes after the last remembered number suffix. The processing will involve reading the latest ACL associated with the corresponding resource. Since every sequential node is unique, the ACL manager won't miss any ACL change. 

On initialization, the ACL manager will first register the watcher and read all existing ACLs.

Finally, we need a way to garbage collect old sequential nodes. This can be done by just removing sequential nodes that are say, more than 15 mins old (assuming every broker would have picked up the ACL changes by then).

Parth-Brahmbhatt replied:
@junrao  Thanks for the explanation and it makes perfect sense. Will update the PR.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+import scala.collection.JavaConverters._
+
+object SimpleAclAuthorizer {
+  //the zookeeper cluster where acls will be stored, defaults to value specified for "zkConnect"
```
Could we document the format of values used in the new ZK paths in the comment?

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+
+    //we still invalidate the cache every hour in case we missed any watch notifications due to re-connections.
+    scheduler.startup()
+    scheduler.schedule("sync-acls", syncAcls, delay = 0l, period = 1l, unit = TimeUnit.HOURS)
```
Hmm, do we need this? It's simpler if we always rely on ZK watchers to propagate the changes. ZKClient will handle all the reconnects for us. If we lose a session, we probably have to read the AclChangedZkPath in case we missed a ZK notification.

Parth-Brahmbhatt replied:
removed and added listener for reconnection handler.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+class SimpleAclAuthorizer extends Authorizer with Logging {
+  private var superUsers: Set[KafkaPrincipal] = Set.empty
+  private var shouldAllowEveryoneIfNoAclIsFound = false
+  private var zkClient: ZkClient = null
```
You can use `_` if you want a value to be initialised to the default one (eg `null`, `false`, `0`, etc.)

Parth-Brahmbhatt replied:
Done.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+  private var superUsers: Set[KafkaPrincipal] = Set.empty
+  private var shouldAllowEveryoneIfNoAclIsFound = false
+  private var zkClient: ZkClient = null
+  private val scheduler: KafkaScheduler = new KafkaScheduler(threads = 1, threadNamePrefix = "authorizer")
```
Type annotation is redundant in cases like this.

Parth-Brahmbhatt replied:
removed the whole scheduler.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+  private var zkClient: ZkClient = null
+  private val scheduler: KafkaScheduler = new KafkaScheduler(threads = 1, threadNamePrefix = "authorizer")
+
+  private val aclCache: scala.collection.mutable.HashMap[Resource, Set[Acl]] = new scala.collection.mutable.HashMap[Resource, Set[Acl]]
```
A less verbose and more idiomatic way to write this:

`import scala.collection._`
...
`private val aclCache = mutable.Map.empty[Resource, Set[Acl]]`

Parth-Brahmbhatt replied:
as soon as I add that import all the other places that expect Set starts complaining that "Set[X] can not be converted to Set[X]" which I am guessing is just scala compiler complaining that they expect the unmutable scala set but with the new import now they are all assumed to be mutable set. Did I mention scala is amazing :-).

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+
+    superUsers = configs.get(SimpleAclAuthorizer.SuperUsersProp) match {
+      case null => Set.empty[KafkaPrincipal]
+      case (str: String) => if (!str.isEmpty) str.split(",").map(s => KafkaPrincipal.fromString(s.trim)).toSet else Set.empty
```
What if `str` is not a `String`? That case is not being handled. Nicer code can be written if you use the scala version of the map (ie `config`). Then `get` will return an `Option` and you can use methods like `filter`, `map`, etc.

That applies to the other code that is reading from the config.

Parth-Brahmbhatt replied:
Handled the case where its not string by just adding another case.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  private def aclMatch(session: Session, operations: Operation, resource: Resource, principal: KafkaPrincipal, host: String, permissionType: PermissionType, acls: Set[Acl]): Boolean = {
+    for (acl: Acl <- acls) {
```
Unnecessary type annotation

Parth-Brahmbhatt replied:
Fixed.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+    for (acl: Acl <- acls) {
+      if (acl.permissionType.equals(permissionType)
+        && (acl.principal.equals(principal) || acl.principal.equals(Acl.WildCardPrincipal))
+        && (operations.equals(acl.operation) || acl.operation.equals(All))
```
In Scala, one should use `==` not `equals`.

Parth-Brahmbhatt replied:
Done.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+        && (operations.equals(acl.operation) || acl.operation.equals(All))
+        && (acl.host.contains(host) || acl.host.contains(Acl.WildCardHost))) {
+        debug(s"operation = $operations on resource = $resource from host = $host is $permissionType.name based on acl = $acl")
+        return true
```
Instead of doing this, you should replace the for comprehension with `find`.

Parth-Brahmbhatt replied:
apart from returning true , given I also want to log the debug statement as a side effect, the version with find looks less redable IMO.

ijuma replied:
Early return inside a closure (and a for comprehension desugars to `foreach`) actually involves throwing a `NonLocalReturnControl` exception. You don't think the following is more readable?

```
 acls.find { acl =>
  acl.permissionType == permissionType
    && (acl.principal == principal || acl.principal == Acl.WildCardPrincipal)
    && (operations == acl.operation || acl.operation == All))
    && (acl.host.contains(host) || acl.host.contains(Acl.WildCardHost))
}.map {
  debug(s"operation = $operations on resource = $resource from host = $host is $permissionType.name based on acl = $acl")
  acl.isDefined
}
```

Parth-Brahmbhatt replied:
the original version is more readable to me, may be its just my java brain. I have assumed scala idioms make it more redable with your approach so changed it as you suggested. Still had to use a return inside the map as acl is not an Option but type ACL.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+
+  def syncAcls(): Unit = {
+    debug("Syncing the acl cache for all resources that are currently in cache ")
+    var resources: collection.Set[Resource] = Set.empty
```
No `var` needed, `inReadLock` returns a value.

Parth-Brahmbhatt replied:
function is deleted given Jun has recommended to assume that zkClient will handle reconnects and guarantee all the watchers are always delivered.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    val aclJson: Option[String] = ZkUtils.readDataMaybeNull(zkClient, toResourcePath(resource))._1
+    val acls: Set[Acl] = aclJson map ((x: String) => Acl.fromJson(x)) getOrElse Set.empty
```
`aclJson.map(Acl.fromJson).getOrElse(Set.empty)`

Parth-Brahmbhatt replied:
Done.

Parth-Brahmbhatt committed:
Adding zkNodeChangeNotificationListener and addressed some comments from Jun and Ismael.

Parth-Brahmbhatt commented:
@junrao Added a generic ZkNodeChangeListener and removed the scheduler thread. Couldn't think of a better name for the class so if you have a better name let me know. Addressed all other comments in this PR.

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,137 @@
+ *
+ * The caller/user of this class should ensure that they use zkClient.subscribeStateChanges and call processAllNotifications
+ * method of this class from ZkStateChangeListener's handleNewSession() method. This is necessary to ensure that if zk session
+ * is terminated and reestablished any missed notification will be processed immeditatly.
```
Typo immeditatly

Parth-Brahmbhatt replied:
Fixed

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,137 @@
+   * @param notifications
+   */
+  private def purgeObsoleteNotifications(now: Long, notifications: Seq[String]) {
+    println("purging notifications" + notifications)
```
Need to either remove those pirintlns or convert them to logging.

Parth-Brahmbhatt replied:
removed debug statement.

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,137 @@
+        import scala.collection.JavaConverters._
+        processNotifications(notifications.asScala.sorted)
+      } catch {
+        case e: Exception => error("Error processing config change:", e)
```
Perhaps we can include path and notifications in the error message.

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+  val AclZkPath = "/kafka-acl"
+
+  //notification node which gets updated with the resource name when acl on a resource is changed.
+  val AclChangedZkPath = "/kafka-acl-changed"
```
We probably should rename this to /kafka-acl-changes.

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+  //notification node which gets updated with the resource name when acl on a resource is changed.
+  val AclChangedZkPath = "/kafka-acl-changed"
+
+  val AclChangedPrefix = "aclChanged"
```
Perhaps rename this to "acl_changes_"?

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+  /**
+   * Guaranteed to be called before any authorize call is made.
+   */
+  override def configure(configs: util.Map[String, _]): Unit = {
```
Probably rename this as javaConfigs. Then we can define the scala one as configs.

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+  private var superUsers = Set.empty[KafkaPrincipal]
+  private var shouldAllowEveryoneIfNoAclIsFound: Boolean = _
+  private var zkClient: ZkClient = _
+  private var aclChangeListener: ZkNodeChangeNotificationListener = _
```
Perhaps it's better to explicitly assign the initial values than using _.

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+    superUsers = configs.get(SimpleAclAuthorizer.SuperUsersProp) match {
+      case null => Set.empty[KafkaPrincipal]
+      case (str: String) => if (!str.isEmpty) str.split(",").map(s => KafkaPrincipal.fromString(s.trim)).toSet else Set.empty
+      case _ => Set.empty
```
To be consistent, use Set.empty[KafkaPrincipal] instead of Set.empty?

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+
+    val zkUrl = config.getOrElse(SimpleAclAuthorizer.ZkUrlProp, kafkaConfig.zkConnect).toString
+
+    zkClient = ZkUtils.createZkClient(zkUrl, Integer2int(kafkaConfig.zkConnectionTimeoutMs), Integer2int(kafkaConfig.zkConnectionTimeoutMs))
```
Perhaps we can allow zkConnectionTimeout and zkSessionTimeout to be configured explicitly for SimpleAuthorizer and default to those in KafkaConfig if not specified.

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+  }
+
+  override def authorize(session: Session, operation: Operation, resource: Resource): Boolean = {
+    val principal: KafkaPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, session.principal.getName)
```
I had a comment in the jira. It seems that in SocketServer, when creating the session object, it's better to create a KafkaPrincipal instead of using KafkaChannel.principal(). The type in KafkaPrincipal should always be USER and the name should be KafkaChannel.principal().getName(). Then, we can just get the KafkaPrincipal from Session and don't need to create a new one.

Parth-Brahmbhatt replied:
I like the idea of changing type of principal in session as KafkaPrincipal. I dont agree with the Type being always user as different authentication schemes may want to change that and different authorizer implementations may actually use the values differently. Does that sound ok?

junrao replied:
Hmm, how do we get the type from the authentication layer? Currently, Authenticator only returns a Principal, which just has a name?

Parth-Brahmbhatt replied:
I was saying we change the type in Authenticator interface to KafkaPrincipal. Then each authenticator implementation can add its own type (SSL can set principalType=Certificate, and kerberos based sasl can set type= User) as long as the authorizer they use can handle different types they will be fine. This will specifically useful when there are multiple listeners.   

I guess we can just do what you are suggesting for now and if we encounter a real use case we can change the type in authenticator at that time. I have made the changes you suggested for now.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+    val principal: KafkaPrincipal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, session.principal.getName)
+    val host = session.host
+
+    trace(s"Authorizer invoked to authorize principal = $principal from host = $host on resource = $resource for operation = $resource")
```
This is useful for things like auditing. Perhaps we can log this in the end together with the decision on wether the operation is granted on not. Also, could we put that into a separate authorization log4j logger (take a look at the request logger in log4j.properties)? That way, if people want auditing, they can just enable the authorization logger.

Parth-Brahmbhatt replied:
Any time we allow or deny we already log the decision and the reasoning behind it, the trace is actually just to mark the entry point. 

Modified log4j but I have set the default level at warn which will disable audit logging by default. Let me know if you want it to be enabled by default.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+    false
+  }
+
+//  private def aclMatch(session: Session, operations: Operation, resource: Resource, principal: KafkaPrincipal, host: String, permissionType: PermissionType, acls: Set[Acl]): Boolean = {
```
Should this be removed?

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+    }
+  }
+
+  def getAclsFromZk(resource: Resource): Set[Acl] = {
```
Can this be private?

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+    acls.toSet[Acl]
+  }
+
+  def loadCache: Set[Acl] = {
```
Can this be private?

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+  }
+
+
+  def toResourcePath(resource: Resource): String = {
```
Can this be private?

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,286 @@
+    SimpleAclAuthorizer.AclZkPath + "/" + resource.resourceType + "/" + resource.name
+  }
+
+  def updateAclChangedFlag(resource: Resource): Unit = {
```
Can this be private?

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/test/scala/unit/kafka/common/ZkNodeChangeNotificationListenerTest.scala:
```diff
@@ -0,0 +1,66 @@
+
+
+/**
+ * Created by pbrahmbhatt on 9/11/15.
```
This needs to be removed.

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/test/scala/unit/kafka/common/ZkNodeChangeNotificationListenerTest.scala:
```diff
@@ -0,0 +1,66 @@
+  def testProcessNotification {
+    val notificationHandler = new NotificationHandler {
+      var notification: String = _
+      var invocationCount: Integer = 0
```
Since these two are accessed from different threads, do they need to be volatile?

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,137 @@
+ * @param changeExpirationMs
+ * @param time
+ */
+class ZkNodeChangeNotificationListener(private val zkClient: ZkClient,
```
Ideally, we want to reuse this class in DynamicConfigManager. If that's too much to do in this jira, could you file a follow-up jira to address this?

Parth-Brahmbhatt replied:
Was planning to do the same. Filed KAFKA-2547

Parth-Brahmbhatt committed:
Addressing more comments from JUN.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,278 @@
+      return true
+
+    //We have some acls defined and they do not specify any allow ACL for the current session, reject request.
+    debug(s"principal = $principal from host = $host is not allowed to perform operation = $operation  on resource = $resource as no explicit acls were defined to allow the operation")
```
Since we return early in a few places, it seems that not all accesses are logged.

Parth-Brahmbhatt replied:
I rechecked to assure it is logged in all cases (superuser, no acls, deny acl match(as part of matching function), allow acl match(as part of matching function),  no acl found).

If we want it to be logged at the end, I can change the logic so it does not return in middle. May be its just me, I think the current way looks cleaner.

Parth-Brahmbhatt replied:
added a consistent audit log message in addition to the other logging statements.

Parth-Brahmbhatt committed:
Redirecting all log statement to authorizer log.

Parth-Brahmbhatt committed:
Changing type of Session's principal from Principal to KafkaPrincipal.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,283 @@
+      val resourceType = ResourceType.fromString(rType)
+      for (resources <- ZkUtils.getChildren(zkClient, resourceType.name)) {
+        for (resourceName <- resources)
+          acls ++= getAcls(Resource(resourceType, resourceName.toString))
```
getAcls should be getAclsFromZk, right? Could we add a unit test to test the initial loading?

Parth-Brahmbhatt replied:
Done. Sorry about missing this in first place. added the test for load cache.

ijuma commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,134 @@
+    import JavaConversions._
+    import scala.collection.mutable
+
+    processNotifications((changes: mutable.Buffer[String]).sorted)
```
Why not import `JavaConverters._` along with other imports at the top of the file and then simply do `changes.asJava` here and delete all local imports?

Parth-Brahmbhatt replied:
Done.

ijuma commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,134 @@
+  private def processNotifications(notifications: Seq[String]) {
+    if (notifications.size > 0) {
+      info(s"Processing notification(s) to $seqNodeRoot")
+      val now = time.milliseconds
```
We normally include `()` for side-effecting changes.

Parth-Brahmbhatt replied:
may be I am missing it but where do you see the side effect?

ijuma replied:
`milliseconds()` returns a different result every time, so it's a side-effecting method (it gets data from the system clock typically). Side-effect free methods always return the same result given the same arguments, so a method that takes no parameters and doesn't return a constant result is a side-effecting method.

Parth-Brahmbhatt replied:
my understanding of what side effect is different than what you described. I have added the () anyways given I don't really think it affects readability one way or another.

ijuma replied:
It is true that the terminology would be clearer if we used pure versus impure functions. Thanks for making the change anyway.

Parth-Brahmbhatt replied:
@ijuma I actually had to remove the bracket for compilation to succeed. With the beackets I get 
"/Users/pbrahmbhatt/repo/kafka/core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:80: Long does not take parameters
      val now = time.milliseconds()"

ijuma replied:
Sorry for this. The reason is that the Scala version of `Time` has a method called `milliseconds` so you can't use `()` at the end. The Java version of `Time` doesn't have this issue.

ijuma commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,134 @@
+   * Process the given list of notifications
+   */
+  private def processNotifications(notifications: Seq[String]) {
+    if (notifications.size > 0) {
```
It's generally better to do `notifications.nonEmpty` because it's O(1) even if the underlying implementation has a O(n) `size` (like scala.List).

Parth-Brahmbhatt replied:
Done

ijuma commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,185 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package unit.kafka.security.auth
```
`unit.` should not be part of package name.

Parth-Brahmbhatt replied:
Done

ijuma commented on core/src/test/scala/unit/kafka/common/ZkNodeChangeNotificationListenerTest.scala:
```diff
@@ -0,0 +1,62 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package unit.kafka.common
```
`unit` should not be part of package name

Parth-Brahmbhatt replied:
Done.

Parth-Brahmbhatt committed:
Added test for load cache. Added consistent audit messages.

Parth-Brahmbhatt committed:
Adding () after milliseconds.

Parth-Brahmbhatt committed:
Removing the () that causes compilation failure.

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,132 @@
+import kafka.utils.{Time, SystemTime, ZkUtils, Logging}
+import org.I0Itec.zkclient.{IZkChildListener, ZkClient}
+import scala.collection.JavaConversions
+import JavaConversions._
```
It seems that we don't need both imports on JavaConversions? import scala.collection.JavaConversions._ seems enough. Also, I think the previous version you had to just do the import inside processAllNotifications() is better since it makes it clear where the implicits are used.

ijuma replied:
@junrao, why not simply use import JavaConverters._? It's the recommended way since it was introduced years ago. The main advantage is that it is both explicit (one has to use `asScala` or `asJava`) and you don't need scoped imports everywhere (which are quite verbose).

junrao replied:
Using explicit JavaConverters will be fine.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,288 @@
+    ).map{ acl: Acl =>
+      debug(s"operation = $operations on resource = $resource from host = $host is $permissionType based on acl = $acl")
+      return true
+    }
```
@ijuma : I actually find the "for-loop" version that Parth wrote earlier easier to understand. map() is supposed to convert one collection to another and it seems it's weird to return from inside a map().

ijuma replied:
I agree that this version looks worse. It's not what I suggested. :) A `getOrElse` after the map is what I would do. Do you still prefer the previous version in that case?

ijuma replied:
To elaborate a bit, `return` is discouraged in Scala because it's not composable which makes it harder to refactor code safely. Once `return` is used, it is no longer safe to extract code into reusable functions. Another problem with `return` is that it uses exceptions for control flow once used inside closures (which are everywhere in Scala) as it is the only way to implement the specified behaviour in the JVM.

Unfortunately, return is used quite a lot in this PR. I'd be willing to submit a PR to this branch that removed all usages of `return` for comparison if there is interest.

junrao replied:
Could you post the exact syntax you had in mind?

ijuma replied:
```
acls.find { acl =>
  acl.permissionType == permissionType
  && (acl.principal == principal || acl.principal == Acl.WildCardPrincipal)
  && (operations == acl.operation || acl.operation == All)
  && (acl.host == host || acl.host == Acl.WildCardHost)
}.map { acl =>
  debug(s"operation = $operations on resource = $resource from host = $host is $permissionType based on acl = $acl")
  true
}.getOrElse(false)
```

or

```
acls.find { acl =>
  acl.permissionType == permissionType
  && (acl.principal == principal || acl.principal == Acl.WildCardPrincipal)
  && (operations == acl.operation || acl.operation == All)
  && (acl.host == host || acl.host == Acl.WildCardHost)
}.fold(false) { acl =>
  debug(s"operation = $operations on resource = $resource from host = $host is $permissionType based on acl = $acl")
  true
}
```

The latter is more concise although we don't use it much in Kafka yet.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,288 @@
+      val resourceTypePath = SimpleAclAuthorizer.AclZkPath + "/" + resourceType.name
+      val resourceNames = ZkUtils.getChildren(zkClient, resourceTypePath)
+      for (resourceName <- resourceNames) {
+        for (resourceName <- resourceNames)
```
Why are we looping through resourceNames twice?

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,218 @@
+
+    simpleAclAuthorizer.addAcls(acls, resource)
+
+    TestUtils.waitUntilTrue(() => simpleAclAuthorizer.getAcls(resource) == acls, "changes not propagated in timeout period.")
```
Could we put addAcls() and waitUntilTrue() in a private method and reuse?

Parth-Brahmbhatt committed:
Got rid of all returns.

Parth-Brahmbhatt commented:
@junrao  @ijuma all comments addressed. There is no return statement in the code now.

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,129 @@
+
+import kafka.utils.{Time, SystemTime, ZkUtils, Logging}
+import org.I0Itec.zkclient.{IZkChildListener, ZkClient}
+import scala.collection.JavaConversions._
```
It would be better to move this inside processAllNotifications() so that we know exactly where the implicits are used. Alternatively, we can use the explicit conversion through JavaConverters as Isamel suggested.

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+    authorized
+  }
+
+  def isEmptyAcl(operation: Operation, resource: Resource, principal: KafkaPrincipal, host: String, acls: Set[Acl]): Option[Boolean] = {
```
Could we just return Boolean here?

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+      val filteredAcls: Set[Acl] = existingAcls.filter((acl: Acl) => !aclsTobeRemoved.contains(acl))
+
+      val path: String = toResourcePath(resource)
+      ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(filteredAcls)))
```
It seems that if filteredAcls is empty, we want to remove the corresponding ACL path. Could we add a unit test for that? Also, we probably only need to update ZK if filteredAcls is different from existingAcls.

Parth-Brahmbhatt replied:
we were not deleting the zookeeper path until an explicit call to remove(resource was made). I have changed that and added unit test. to read the zookeeper path I made the method toResourcePath public.

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,218 @@
+    assertEquals(acls, simpleAclAuthorizer.getAcls(resource))
+  }
+
+  private def addAclAndVerify(acls: Set[Acl]) {
```
Could we make this more general to sth like the following? Then we can remove all individual usage of waitUntilTrue().

// return the new acl set after applying the changes to the original acl.
changeAclAndVerify(originalAcl: Set[Acl], addedAcl: Set[Acl], removedAcl: Set[Acl]): Set[Acl]

Parth-Brahmbhatt replied:
Done

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+
+  object ZkStateChangeListener extends IZkStateListener {
+
+    var reconnection = false
```
Is this used anywhere?

Parth-Brahmbhatt replied:
In the initialize we register this zkClient so in case a reconnection happens zkClient will invoke this method and as part handling a new session it will process any missed notifications during the reconnection.

junrao replied:
@ijuma has a valid point. It doesn't seem reconnection is ever used? The handling of the reconnection logic is all in handleNewSession().

Parth-Brahmbhatt replied:
sorry i though he was referring to ZkStateChangeListener object itself. removed.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+  }
+
+  def isSuperUser(operation: Operation, resource: Resource, principal: KafkaPrincipal, host: String): Boolean = {
+    superUsers.find (user => user == principal).map {
```
`exists` is probably better than `find` here as you are not using the value. Maybe something like

```
if (superUsers.exists(_ == principal)) {
  authorizerLogger.debug(...)
  true
}
else false
```

Parth-Brahmbhatt replied:
Done.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+      val existingAcls: Set[Acl] = getAcls(resource)
+      val filteredAcls: Set[Acl] = existingAcls.filter((acl: Acl) => !aclsTobeRemoved.contains(acl))
+
+      val path: String = toResourcePath(resource)
```
A number of typing annotations that could be removed as they are easily inferred (`String`, `Set[Acl]`, `Acl`). It's OK to keep them if you think they help readability (as opposed to just Java habits ;)).

Parth-Brahmbhatt replied:
java habits, nothing really to do with readability :-). Guess it's going to take me sometime before I stop doing this.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+  private var zkClient: ZkClient = null
+  private var aclChangeListener: ZkNodeChangeNotificationListener = null
+
+  private val aclCache: scala.collection.mutable.HashMap[Resource, Set[Acl]] = new scala.collection.mutable.HashMap[Resource, Set[Acl]]
```
Remove type annotation on the left-hand side? Also, you can replace `HashMap` with `Map` which is the static factory method.

Parth-Brahmbhatt replied:
Removed the type. Trying to convert to Map actually gives compilation error.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+    for (aclSet <- aclSets) {
+      acls ++= aclSet.filter(acl => acl.principal == principal)
+    }
+    acls
```
`aclCache.values.flatMap(_.filter(_.principal == principal)).toSet` should do the same as all of the lines above. The following is a bit more readable though:

`aclCache.values.flatMap(aclSet => aclSet.filter(_.principal == principal)).toSet`

Parth-Brahmbhatt replied:
I actually realized one thing, given acl class itself does not have resource as acls are suppose to be attached to a resource, this method is pretty useless unless it returns a Map[Resource, Set[Acl]].  I have made the changes to both the interface and the implementation to reflect that.

Parth-Brahmbhatt committed:
Changing authozier interface so getacl for principal returns a map(resource -> set[Acls]) and some other styling comments.

junrao commented on core/src/main/scala/kafka/security/auth/Authorizer.scala:
```diff
@@ -74,8 +74,8 @@ trait Authorizer extends Configurable {
    * get the acls for this principal.
    * @param principal
-   * @return empty set if no acls exist for this principal, otherwise the acls for the principal.
+   * @return empty set if no acls exist for this principal, otherwise a map of resource -> acls for the principal.
```
empty set => empty map

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,281 @@
+      val aclNeedsRemoval = (existingAcls != filteredAcls)
+      if(aclNeedsRemoval) {
+        val path: String = toResourcePath(resource)
+        if(filteredAcls.nonEmpty)
```
space after if

Parth-Brahmbhatt replied:
Done.

Parth-Brahmbhatt committed:
Styling changes.

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,129 @@
+  /**
+   * create seqNodeRoot and begin watching for any new children nodes.
+   */
+  def startup() {
```
startup() implies this class will have internal threads, but it doesn't. Would it be better to rename this to init()?

Parth-Brahmbhatt replied:
done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+    superUsers = javaConfigs.get(SimpleAclAuthorizer.SuperUsersProp) match {
+      case null => Set.empty[KafkaPrincipal]
+      case (str: String) => if (!str.isEmpty) str.split(",").map(s => KafkaPrincipal.fromString(s.trim)).toSet else Set.empty[KafkaPrincipal]
+      case _ => Set.empty
```
To be consistent with the line in 86, Set.empty => Set.empty[KafkaPrincipal]?

Parth-Brahmbhatt replied:
done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+
+    zkClient.subscribeStateChanges(ZkStateChangeListener)
+
+    loadCache
```
The ordering can be a bit subtle here. In order not to miss a session expiration event, we should probably use the following ordering.

  zkClient = ZkUtils.createZkClient(zkUrl, zkConnectionTimeoutMs, zkSessionTimeOutMs)
    ZkUtils.makeSurePersistentPathExists(zkClient, SimpleAclAuthorizer.AclZkPath)

```
loadCache

zkClient.subscribeStateChanges(ZkStateChangeListener)

aclChangeListener = new ZkNodeChangeNotificationListener(zkClient, SimpleAclAuthorizer.AclChangedZkPath, SimpleAclAuthorizer.AclChangedPrefix, AclChangedNotificaitonHandler)
aclChangeListener.startup()
```

Parth-Brahmbhatt replied:
done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+      val resourceTypePath = SimpleAclAuthorizer.AclZkPath + "/" + resourceType.name
+      val resourceNames = ZkUtils.getChildren(zkClient, resourceTypePath)
+      for (resourceName <- resourceNames)
+        acls ++= getAclsFromZk(Resource(resourceType, resourceName.toString))
```
We should add the Acl to aclCache, right? Could we add a unit test to cover this?

Parth-Brahmbhatt replied:
Originally I did that but I figured given we havent added add,remove APIs in reality add/remove calls will be made from CLI so modifying cache did not make much sense. Also it seemed cleaner to just have a single path (notification) that updated the cache and also made it easy to test. 

I have made changes to update the cache as you suggested but I don't see given the same state can be modified by 2 components there is any easy way to add unit test to ensure which component actually made the state change.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,279 @@
+  object AclChangedNotificaitonHandler extends NotificationHandler {
+
+    override def processNotification(notificationMessage: Option[String]): Unit = {
+      val resource: Resource = Resource.fromString(notificationMessage.get)
```
It seems that we expect notificationMessage to be non-empty. So, perhaps we can change the type of processNotifications to be just String and in ZkNodeChangeNotificationListener.processNotifications(), log a warning (with the path) if the data read from ZK is empty.

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,227 @@
+
+    val authorizer = new SimpleAclAuthorizer
+    authorizer.configure(config.originals)
+    assertEquals(acls, simpleAclAuthorizer.getAcls(resource))
```
simpleAclAuthorizer should be authorizer? Also, we should add a loadCache() tests where all the nodes in /acl_changes are gone.

Parth-Brahmbhatt replied:
Done.

ijuma commented:
@Parth-Brahmbhatt I did a PR to your branch (https://github.com/Parth-Brahmbhatt/kafka/pull/1) with some suggested code readability improvements. If you agree you can merge it to your branch.

Parth-Brahmbhatt committed:
Adding acls to cache as soon as acls are modified rather then relying on notification to be processed.

Parth-Brahmbhatt committed:
Redability changes suggested by Ismael.

Parth-Brahmbhatt commented:
@ijuma changes to ZkNodeChangeNotificationListener can not be accepted as @junrao wants the import to be local.

Other than that  have incorporated your proposed changes.

ijuma commented:
@Parth-Brahmbhatt, Thanks for incorporating the changes. My understanding is that @junrao wanted the import to be local if `JavaConversions` was used. With `JavaConverters`, this is not needed because you have the `asScala` and `asJava` to show you that a conversion is happening (unlike the `JavaConversions` case where everything happens silently).

Parth-Brahmbhatt committed:
extracting a local import to top level

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+    }.toMap
+  }
+
+  private def loadCache() : Unit = {
```
This can just be 
private def loadCache() {
...
}

There are a few other places that we can get rid of ": unit =".

Parth-Brahmbhatt replied:
Done

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+      val resourceTypePath = SimpleAclAuthorizer.AclZkPath + "/" + resourceType.name
+      val resourceNames = ZkUtils.getChildren(zkClient, resourceTypePath)
+      for (resourceName <- resourceNames) {
+        acls ++= getAclsFromZk(Resource(resourceType, resourceName.toString))
```
We shouldn't be accumulating in acls, right?

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,228 @@
+    val acl1 = new Acl(user1, Allow, host1, Read)
+    val acl2 = new Acl(user1, Allow, host1, Write)
+    val acls = Set[Acl](acl1, acl2)
+    simpleAclAuthorizer.addAcls(acls, resource)
```
Could we set up two resources with ACL to cover the new issue identified in loadCache()?

Parth-Brahmbhatt replied:
Yes, done.

junrao commented on core/src/main/scala/kafka/common/ZkNodeChangeNotificationListener.scala:
```diff
@@ -0,0 +1,130 @@
+    override def handleChildChange(path: String, notifications: java.util.List[String]) {
+      try {
+        import scala.collection.JavaConverters._
+        if(notifications != null)
```
space after if

Parth-Brahmbhatt replied:
Done.

junrao commented:
Also, for JavaConverters, it's ok to import it globally since we are using asScala explicitly.

Parth-Brahmbhatt committed:
BUG FIX: load cache should not have the acls appended.

Parth-Brahmbhatt commented:
All comments addressed.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+  /**
+   * Guaranteed to be called before any authorize call is made.
+   */
+  override def configure(javaConfigs: util.Map[String, _]): Unit = {
```
remove ": Unit ="

Parth-Brahmbhatt replied:
Done.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+    }.toMap
+  }
+
+  private def loadCache() = {
```
If a method has no return value, we need to define it as the following w/o =. Otherwise, it will pick up the value in the last statement as the return value. Ditto in a few other places.

def loadCache() {
}

Parth-Brahmbhatt replied:
Done.

Parth-Brahmbhatt committed:
more styling.

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+    SimpleAclAuthorizer.AclZkPath + "/" + resource.resourceType + "/" + resource.name
+  }
+
+  private def logAuditMessage(principal: KafkaPrincipal, authorized: Boolean, operation: Operation, resource: Resource, host: String) = {
```
remove =

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+    authorizerLogger.debug(s"Principal = $principal is $permissionType Operation = $operation from host = $host on resource = $resource")
+  }
+
+  private def updateAclChangedFlag(resource: Resource) = {
```
remove =

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+
+  object AclChangedNotificaitonHandler extends NotificationHandler {
+
+    override def processNotification(notificationMessage: String) = {
```
remove =

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+      aclChangeListener.processAllNotifications
+    }
+
+    override def handleSessionEstablishmentError(error: Throwable) = {
```
remove =

junrao commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -0,0 +1,284 @@
+      fatal("Could not establish session with zookeeper", error)
+    }
+
+    override def handleStateChanged(state: KeeperState) = {
```
remove =

junrao commented on core/src/test/scala/unit/kafka/common/ZkNodeChangeNotificationListenerTest.scala:
```diff
@@ -0,0 +1,61 @@
+  override def generateConfigs() = List(KafkaConfig.fromProps(TestUtils.createBrokerConfig(0, zkConnect)))
+
+  @Test
+  def testProcessNotification {
```
testProcessNotification()

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testTopicAcl(): Unit = {
```
def testTopicAcl() {

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testDenyTakesPrecedence(): Unit = {
```
def testDenyTakesPrecedence() {

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testAllowAllAccess(): Unit = {
```
def testAllowAllAccess() {

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testSuperUserHasAccess(): Unit = {
```
def testSuperUserHasAccess() {

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testNoAclFound(): Unit = {
```
def testNoAclFound() {

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testNoAclFoundOverride(): Unit = {
```
def testNoAclFoundOverride() {

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testAclManagementAPIs(): Unit = {
```
def testAclManagementAPIs() {

junrao commented on core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  @Test
+  def testLoadCache(): Unit = {
```
def testLoadCache() {

Parth-Brahmbhatt committed:
Same styling changes applied to test cases too.

junrao commented:
Thanks a lot for the patch. LGTM.

ijuma commented:
Thanks Parth!
